src/Symfony/Component/TypeInfo/Type/IntersectionType.php

```diff
@@ -11,24 +11,51 @@
 
 namespace Symfony\Component\TypeInfo\Type;
 
+use Symfony\Component\TypeInfo\Exception\InvalidArgumentException;
 use Symfony\Component\TypeInfo\Exception\LogicException;
 use Symfony\Component\TypeInfo\Type;
+use Symfony\Component\TypeInfo\TypeIdentifier;
 
 /**
  * @author Mathias Arlaud <mathias.arlaud@gmail.com>
  * @author Baptiste Leduc <baptiste.leduc@gmail.com>
  *
- * @template T of Type
+ * @template T of ObjectType|GenericType<ObjectType>|CollectionType<GenericType<ObjectType>>
  *
  * @experimental
  */
-final class IntersectionType extends Type
+final class IntersectionType extends Type implements CompositeTypeInterface
 {
     /**
      * @use CompositeTypeTrait<T>
      */
     use CompositeTypeTrait;
 
+    /**
+     * @param list<T> $types
+     */
+    public function __construct(Type ...$types)
+    {
+        if (\count($types) < 2) {
+            throw new InvalidArgumentException(\sprintf('"%s" expects at least 2 types.', self::class));
+        }
+        // Only accept non-composite object types, except the builtin 'object'
```

Comment:
what about `mixed` ?

Reply:
Fully agree about the interface, not sure why I finally decided to add it. Removed and re-added the methods to the `Type` class (might have messed the diff for now, will fix it, sorry).

`mixed` is not allowed as a member of an intersection. `mixed`, `never` and `void` are *standalone* types in PHP, treated very differently from the others: they cannot be used in unions or intersections.

I agree we want to stay as flexible as possible, but we also want to avoid nightmarish edge cases like this current behaviour:

```php
$type = Type::generic(Type::mixed(), Type::int());
dump($type->isNullable()); // true

$nonNullable = $type->asNonNullable();
dump($nonNullable->isNullable()); // true
```